### PR TITLE
Optionally avoid global constructors

### PR DESCRIPTION
**Pull request by ksperling-apple**
Introduce a chip::Global wrapper for managing global objects with non-trivial constructors / destructors.

The behavior of the wrapper can be configured via two `CHIP_CONFIG_GLOBALS_*` options in the platform config header (the default is current standard C++ behavior). By setting `CHIP_CONFIG_GLOBALS_LAZY_INIT = 1` globals can instead be constructed lazily (note this is done in a thread-unsafe way as synchronization is assumed to be external).

Adopt chip::Global in the Darwin platform layer and for key core singletons and enable CHIP_CONFIG_GLOBALS_LAZY_INIT = 1 on Darwin.

### CODE REVIEW

File: src/access/AccessControl.cpp [hunk 1]
```diff
@@ -629,9 +635,17 @@ void AccessControl::NotifyEntryChanged(const SubjectDescriptor * subjectDescript
...
+    static AccessControl sInstance;
+    return &sInstance;
+}
+
+AccessControl * globalAccessControl = nullptr;
+
 AccessControl & GetAccessControl()
 {
-    return *globalAccessControl;
+    return *(globalAccessControl ?: GetDefaultAccessControl());
```
Comment (Damian-Nordic) [inline src/access/AccessControl.cpp:648]: AFAIK, omitting the argument between ? and : is a GNU extension. Please follow the C++ standard.

File: src/access/AccessControl.cpp [hunk 2]
```diff
@@ -174,8 +171,17 @@ char GetPrivilegeStringForLogging(Privilege privilege)
 
 } // namespace
 
-AccessControl::Entry::Delegate AccessControl::Entry::mDefaultDelegate;
-AccessControl::EntryIterator::Delegate AccessControl::EntryIterator::mDefaultDelegate;
+AccessControl::Entry::Delegate & AccessControl::Entry::GetDefaultDelegate()
+{
+    static Delegate instance;
```
Comment (Damian-Nordic) [inline src/access/AccessControl.cpp:176]: Note that this has a code size and runtime penalty compared to the previous version because C++ requires that static local variable initialization is thread-safe: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables. Do you know if the increase reported by bloatcheck if because of that or because of the changes in Optional class? In the first case, you could consider introducing a helper class for controlling initialization of global objects on different platforms.

```c++
#if CHIP_CONFIG_GLOBAL_LAZY_INITIALIZATION

template <class T>
class Global
{
public:
    T& operator*()
    {
      if (!mObject.HasValue())
      {
        mObject.Emplace();
      }
      
      return mObject;
    }

private:
    Optional<T> mObject;
};

#else

template <class T>
class Global
{
public:
    T& operator*() { return *mObject; }

private:
    T* mObject;
};
#endif
```

File: src/lib/core/Optional.h [hunk 3]
```diff
@@ -37,77 +37,119 @@ struct NullOptionalType
...
+
+} // namespace detail
+
 /**
  * Pairs an object with a boolean value to determine if the object value
  * is actually valid or not.
  */
 template <class T>
-class Optional
+class Optional : protected detail::OptionalBase<T>
```
Comment (andy31415) [inline src/lib/core/Optional.h:87]: Could we directly wrap a `std::optional` instead now that we are C++17 and also mark this deprecated for std::optional directly?

File: src/lib/core/Optional.h [hunk 4]
```diff
@@ -37,77 +37,119 @@ struct NullOptionalType
...
+struct OptionalBase
+{
+    bool mHasValue;
+    union
+    {
+        char mEmpty;
+        T mValue;
+    };
+
+    constexpr OptionalBase() : mHasValue(false), mEmpty(0) {}
```
Comment (Damian-Nordic) [inline src/lib/core/Optional.h:53]: Actually, zero-initialization of `mEmpty` might be causing the code size increase. How about using `std::aligned_storage_t<sizeof(T), alignof(T)> mValue` instead? Also, I wonder if you actually need both versions as if a type is trivially desctructible then `~T()` is probably optimized out, right?

File: src/access/AccessControl.cpp (hunk 2 above)
Comment (ksperling-apple) [reply src/access/AccessControl.cpp:176]: I'll split this up into separate commits to see if the Optional change is affecting code size negatively by itself (I've also fixed it to use a 0-sized `std::monostate` instead of a dummy char for the empty case).

A helper class like what you've sketched out might be the way to go for making globals optionally lazy.

File: src/lib/core/Optional.h (hunk 4 above)
Comment (ksperling-apple) [reply src/lib/core/Optional.h:53]: I think using `std::monostate` (aka an empty struct) instead of `char` here should help since it will make the compiler happy that one of the variants is being initialized, without actually initializing any memory at runtime.

In my experiments I'm not seeing these destructors optimized away, but it may be something in my setup. E.g. I'm seeing global initializers emitted to register the destructor for:
```
static Optional<System::Clock::Timeout> activeThresholdTimeOverride  = NullOptional;
```

File: src/lib/core/Optional.h (hunk 3 above)
Comment (ksperling-apple) [reply src/lib/core/Optional.h:87]: We could, but it would be a more substantial change. Maybe the migration to std::optional is something we should tackle separately?

File: src/access/AccessControl.cpp (hunk 2 above)
Comment (ksperling-apple) [reply src/access/AccessControl.cpp]: I've implemented something like this approach now with a `chip::Global` helper.

File: src/lib/core/Optional.h (hunk 3 above)
Comment (ksperling-apple) [reply src/lib/core/Optional.h]: Turns out with correct optimization settings the destructors of chip::Optional get optimized away fine as is even if they're not technically trivial, so I'm leaving it as is.

File: src/lib/core/Optional.h (hunk 4 above)
Comment (ksperling-apple) [reply src/lib/core/Optional.h]: Leaving chip::Optional as is for now

File: src/platform/Darwin/CHIPPlatformConfig.h [hunk 5]
```diff
@@ -27,6 +27,12 @@
 
 #define CHIP_CONFIG_ABORT() abort()
 
+#include <os/trace_base.h> // for __dso_handle
+extern "C" int __cxa_atexit(void (*f)(void *), void * p, void * d);
+#define CHIP_CXA_ATEXIT(f, p) __cxa_atexit((f), (p), &__dso_handle)
```
Comment (Damian-Nordic) [inline src/platform/Darwin/CHIPPlatformConfig.h:32]: Perhaps, this macro could just be called `CHIP_ATEXIT`  as some platforms may not provide `__cxa_atexit`?

**Review by Damian-Nordic — APPROVED** [review]

File: src/access/AccessControl.h [hunk 6]
```diff
@@ -104,15 +105,15 @@ class AccessControl
 
         Entry() = default;
 
-        Entry(Entry && other) : mDelegate(other.mDelegate) { other.mDelegate = &mDefaultDelegate; }
+        Entry(Entry && other) : mDelegate(other.mDelegate) { other.mDelegate = &*mDefaultDelegate; }
```
Comment (tcarmelveilleux) [inline src/access/AccessControl.h:108]: This is really not obvious syntax. Most average developers do not expect this.

**Comment by mspang** [comment]
Can you describe the problem this is aiming to fix?

Wouldn't it be a lot cleaner to put them inside another object? e.g. something like #19784

You can also get lazy initialization by making them function scoped static variables.

File: build/config/compiler/compiler.gni [hunk 7]
```diff
@@ -21,7 +21,8 @@ declare_args() {
 
   # Optimize debug builds with -Og.
   optimize_debug = current_os == "freertos" || current_os == "zephyr" ||
-                   current_os == "mbed" || current_os == "cmsis-rtos"
+                   current_os == "mbed" || current_os == "cmsis-rtos" ||
+                   current_os == "mac" || current_os == "ios"
```
Comment (bzbarsky-apple) [inline build/config/compiler/compiler.gni:25]: Why this change?  `-Og` is decent, but it still makes debugging a bit of a pain compared to not optimizing at all...

File: src/access/AccessControl.h (hunk 6 above)
Comment (bzbarsky-apple) [inline src/access/AccessControl.h:108]: Maybe Global should have an AsPointer method?

File: src/credentials/attestation_verifier/DeviceAttestationVerifier.h [hunk 8]
```diff
@@ -226,20 +226,17 @@ class WellKnownKeysTrustStore
 class ArrayAttestationTrustStore : public AttestationTrustStore
 {
 public:
-    ArrayAttestationTrustStore(const ByteSpan * derCerts, size_t numCerts) : mDerCerts(derCerts), mNumCerts(numCerts) {}
+    ArrayAttestationTrustStore(const ByteSpan * const * derCerts, size_t numCerts) : mDerCerts(derCerts), mNumCerts(numCerts) {}
```
Comment (bzbarsky-apple) [inline src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: So this now asks for an array of _pointers_ to ByteSpan, and the actual ByteSpans have to live somewhere outside the array.

Why is this change needed?  If we do need to make it, that might be better as a separate PR.

File: src/darwin/Framework/CHIP/MTRControllerAccessControl.mm [hunk 9]
```diff
@@ -68,17 +69,23 @@ CHIP_ERROR Check(
...
+
 @implementation MTRControllerAccessControl
 
 + (void)init
 {
-    static dispatch_once_t onceToken;
-    dispatch_once(&onceToken, ^{
-        GetAccessControl().Init(&gDelegate, gDeviceTypeResolver);
-    });
+    (void) *gControllerAccessControl;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRControllerAccessControl.mm:88]: The side-effect from `*` here is a bit weird.  It would be better if there were just an explicit `Emplace()` or `EnsureValue` or something method.

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -1773,7 +1773,7 @@
 				GCC_C_LANGUAGE_STANDARD = gnu11;
 				GCC_DYNAMIC_NO_PIC = NO;
 				GCC_NO_COMMON_BLOCKS = YES;
-				GCC_OPTIMIZATION_LEVEL = 0;
+				GCC_OPTIMIZATION_LEVEL = g;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/Matter.xcodeproj/project.pbxproj:1776]: This is the same "enable -Og for debug builds" thing as above, right?  Why do we want that?

File: src/lib/core/Global.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ * The contained object of type T is default constructed, possibly lazily.
+ *
+ * This class is generally NOT thread-safe; external synchronization is required.
+ */
+template <class T>
+class Global
+{
+public:
+    /// Returns the global object, initializing it if necessary.
+    T & operator*() { return get(); }
```
Comment (bzbarsky-apple) [inline src/lib/core/Global.h:39]: Worth also having a `operator T & () { return get(); }` and the const version of that?  Would allow getting rid of a bunch of `*` this PR adds.

**Review by bzbarsky-apple — COMMENTED** [review]
I recall a previous version of this enabled some compile flags on darwin that would fail for global constructors.  Did those go away?

**Comment by bzbarsky-apple** [comment]
> You can also get lazy initialization by making them function scoped static variables (this way is also thread safe).

@mspang See previous discussion in this PR, which GitHub helpfully hides by default.

File: src/lib/core/Global.h [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    T & get() { return mValue; }
+
+#if CHIP_CONFIG_GLOBALS_NO_DESTRUCT
+public:
+    // For not-trivially-destructible T, hiding it inside a union means the destructor
+    // won't get called automatically, however our own default destructor will be implicitly
+    // deleted. So Global<T> is technically not trivially-destructible, however the compiler
+    // tends to be able to optimize away the empty destructor call in practice. Getting around
+    // this cleanly requires using the "storage" approach used in the lazy variant, however
+    // this would require std::construct_at from C++20 to get constexpr initialization.
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/lib/core/Global.h:84]: There's no constexpr way to do placement new prior to C++20, so in my experiments it ended up losing constant initialization in cases that would otherwise have had it.

File: src/platform/Darwin/CHIPPlatformConfig.h (hunk 5 above)
Comment (ksperling-apple) [reply src/platform/Darwin/CHIPPlatformConfig.h:32]: Regular atexit doesn't have the `void *` parameter so it's harder to use for this. Platforms don't need to provide this unless they use the lazy option without no_destruct.

Not sure the name matters too much -- the semantics are expected to be like `__cxa_atexit` so I named it based on that.

File: src/credentials/attestation_verifier/DeviceAttestationVerifier.h (hunk 8 above)
Comment (ksperling-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: The way we're currently creating the array for the single use case in the SDK is like this:
```
static const ByteSpan kTestPaaRoots[] = {
    TestCerts::sTestCert_PAA_FFF1_Cert,
    TestCerts::sTestCert_PAA_NoVID_Cert,
};
```
This is not constinit, because the ByteSpan objects live elsewhere and have to be copied.

Defining a named constinit ByteSpan object for each cert and then collecting them into a list seems like an ergonomic way to use this, so making the array contain a pointer to the ByteSpan seems better than e.g. creating the ByteSpan directly inside the roots array.

File: src/access/AccessControl.h (hunk 6 above)
Comment (ksperling-apple) [reply src/access/AccessControl.h:108]: It's standard for things like smart pointers, std::optional etc...

I'm not too fussed either way... we can make `get()` public for use cases like this one, or we can even get rid of the operator overloads entirely and use `get()` everywhere to make it obvious at the call site that there's potentially more happening than just a plain variable access or pointer dereference.

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj (hunk 10 above)
Comment (ksperling-apple) [reply src/darwin/Framework/Matter.xcodeproj/project.pbxproj:1776]: I can take these back out... It's "the same" but not redundant because the setting doesn't propagate between the build systems.

File: src/lib/core/Global.h (hunk 11 above)
Comment (ksperling-apple) [reply src/lib/core/Global.h:39]: I think that's too "magical"... I'd like to see something at the call site that makes it obvious or at least hints that there is more than just a variable access going on.

File: src/credentials/attestation_verifier/DeviceAttestationVerifier.h (hunk 8 above)
Comment (bzbarsky-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: > because the ByteSpan objects live elsewhere and have to be copied.

More precisely, they are extern and you can't have an extern constexpr, right?

But then what we are really doing is replacing non-const init with linker relocations to fix up the pointers, which is about the same in terms of startup time....

Comment (ksperling-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: It's easy to validate / enforce mechanically that there are no initializers though (via `-no_inits` and/or `constinit`), whereas seeing whether or not an initializer does a lot of work is difficult.

Comment (bzbarsky-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: OK.  So what I am wondering, @ksperling-apple is now that we are c++17 whether we can make the "public" bits in `src/credentials/tests/CHIPAttCert_test_vectors.h` be `inline constexpr` so that we don't have to worry about the extern thing at all....

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj (hunk 10 above)
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/Matter.xcodeproj/project.pbxproj]: Removed those commits

File: src/access/AccessControl.h (hunk 6 above)
Comment (ksperling-apple) [reply src/access/AccessControl.h]: I have removed the `operator*` overload and made `get()` public, this should make call sites more obvious.

File: build/config/compiler/compiler.gni (hunk 7 above)
Comment (ksperling-apple) [reply build/config/compiler/compiler.gni]: I've removed that commit from the PR now.

File: src/access/AccessControl.h (hunk 6 above)
Comment (ksperling-apple) [reply src/access/AccessControl.h]: This is now `&mDefaultDelegate.get()` which seems fine.

File: src/credentials/attestation_verifier/DeviceAttestationVerifier.h (hunk 8 above)
Comment (ksperling-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: Hm, I guess that's an option, but having these large constants defined extern and have them in only one translation unit seems generally the right thing. We do seem to have a bunch of other `constexpr` constants that should actually be `inline constexpr`. I'll put up a separate PR for that.

File: src/darwin/Framework/CHIP/MTRControllerAccessControl.mm (hunk 9 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRControllerAccessControl.mm:88]: Using `get()` instead of `operator*` now.

File: src/credentials/attestation_verifier/DeviceAttestationVerifier.h (hunk 8 above)
Comment (bzbarsky-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h:229]: @ksperling-apple I think the issue is that ArrayAttestationTrustStore is public API.  So changing it if we don't have to is not great....

I agree that having the large constants be extern is generally the right thing; just trying to figure out a way to deal with this without changing public APIs.

Comment (ksperling-apple) [reply src/credentials/attestation_verifier/DeviceAttestationVerifier.h]: Fixed by moving `kTestAttestationTrustStoreRoots` (formerly `kTestPaaRoots`) into `CHIPAttCert_test_vectors.cpp` where the certificate spans themselves are defined.

File: src/lib/core/Global.h [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+ * The contained object of type T is default constructed, possibly lazily.
+ *
+ * This class is generally NOT thread-safe; external synchronization is required.
+ */
+template <class T>
+class Global
+{
+public:
+    /// Returns the global object, initializing it if necessary.
+    /// NOT threads-safe, external synchronization is required.
```
Comment (bzbarsky-apple) [inline src/lib/core/Global.h:39]:
```suggestion
    /// NOT thread-safe, external synchronization is required.
```

**Review by bzbarsky-apple — APPROVED** [review]
